### ISSUE TITLE
Build fails against PostgreSQL 9.6

### ISSUE THREAD

**Issue by devrimgunduz** [ossc-db/pg_bulkload#20]
Hi,

Getting this against PostgreSQL 9.6, which will be released on Thursday. Can we please get a new version that includes the fix soon?

Thanks!

```
+ make USE_PGXS=1 -j3
pg_btree.c:52:2: error: #error unsupported PostgreSQL version
 #error unsupported PostgreSQL version
  ^
pg_btree.c:107:42: error: unknown type name 'BTWriteState'
 static void _bt_mergeload(Spooler *self, BTWriteState *wstate, BTSpool *btspool,
                                          ^
pg_btree.c: In function 'IndexSpoolBegin':
pg_btree.c:207:33: error: 'BTREE_AM_OID' undeclared (first use in this function)
    indices[i]->rd_rel->relam == BTREE_AM_OID)
                                 ^
pg_btree.c:207:33: note: each undeclared identifier is reported only once for each function it appears in
pg_btree.c:222:13: error: dereferencing pointer to incomplete type
    spools[i]->isunique = indices[i]->rd_index->indisunique;
             ^
pg_btree.c: In function 'IndexSpoolInsert':
pg_btree.c:290:12: warning: variable 'heapRelation' set but not used [-Wunused-but-set-variable]
  Relation  heapRelation;
            ^
pg_btree.c: In function '_bt_mergebuild':
pg_btree.c:371:2: error: unknown type name 'BTWriteState'
  BTWriteState wstate;
  ^
pg_btree.c:377:31: error: dereferencing pointer to incomplete type
  tuplesort_performsort(btspool->sortstate);
                               ^
pg_btree.c:379:8: error: request for member 'index' in something not a structure or union
  wstate.index = btspool->index;
        ^
pg_btree.c:379:24: error: dereferencing pointer to incomplete type
  wstate.index = btspool->index;
                        ^
pg_btree.c:387:8: error: request for member 'btws_use_wal' in something not a structure or union
  wstate.btws_use_wal = self->use_wal &&
        ^
In file included from /usr/pgsql-9.6/include/server/storage/lmgr.h:20:0,
                 from pg_btree.c:21:
pg_btree.c:388:46: error: request for member 'index' in something not a structure or union
   XLogIsNeeded() && !RELATION_IS_LOCAL(wstate.index);
                                              ^
/usr/pgsql-9.6/include/server/utils/rel.h:479:4: note: in definition of macro 'RELATION_IS_LOCAL'
  ((relation)->rd_islocaltemp || \
    ^
pg_btree.c:388:46: error: request for member 'index' in something not a structure or union
   XLogIsNeeded() && !RELATION_IS_LOCAL(wstate.index);
                                              ^
/usr/pgsql-9.6/include/server/utils/rel.h:480:4: note: in definition of macro 'RELATION_IS_LOCAL'
   (relation)->rd_createSubid != InvalidSubTransactionId)
    ^
pg_btree.c:395:8: error: request for member 'btws_pages_alloced' in something not a structure or union
  wstate.btws_pages_alloced = BTREE_METAPAGE + 1;
        ^
pg_btree.c:396:8: error: request for member 'btws_pages_written' in something not a structure or union
  wstate.btws_pages_written = 0;
        ^
pg_btree.c:397:8: error: request for member 'btws_zeropage' in something not a structure or union
  wstate.btws_zeropage = NULL; /* until needed */
        ^
pg_btree.c:404:21: error: request for member 'index' in something not a structure or union
  LockRelation(wstate.index, AccessExclusiveLock);
                     ^
pg_btree.c:405:29: error: request for member 'index' in something not a structure or union
  FlushRelationBuffers(wstate.index);
                             ^
pg_btree.c:408:38: error: request for member 'index' in something not a structure or union
  merge = BTReaderInit(&reader, wstate.index);
                                      ^
In file included from /usr/pgsql-9.6/include/server/postgres.h:48:0,
                 from ../include/pg_bulkload.h:9,
                 from pg_btree.c:11:
pg_btree.c:411:33: error: request for member 'index' in something not a structure or union
   RelationGetRelationName(wstate.index),
                                 ^
/usr/pgsql-9.6/include/server/utils/elog.h:202:23: note: in definition of macro 'elog'
   elog_finish(elevel, __VA_ARGS__); \
/usr/pgsql-9.6/include/server/utils/rel.h:392:3: note: in expansion of macro  NameStr'
  (NameStr((relation)->rd_rel->relname))
   ^
pg_btree.c:411:3: note: in expansion of macro 'RelationGetRelationName'
   RelationGetRelationName(wstate.index),
   ^
pg_btree.c:413:9: error: request for member 'btws_use_wal' in something not a structure or union
   wstate.btws_use_wal ? "with" : "without");
         ^
/usr/pgsql-9.6/include/server/utils/elog.h:202:23: note: in definition of macro 'elog'
   elog_finish(elevel, __VA_ARGS__); \
                       ^
pg_btree.c:416:34: error: request for member 'index' in something not a structure or union
  RelationSetNewRelfilenode(wstate.index, InvalidTransactionId);
                                  ^
pg_btree.c:416:2: error: too few arguments to function 'RelationSetNewRelfilenode'
  RelationSetNewRelfilenode(wstate.index, InvalidTransactionId);
  ^
In file included from /usr/pgsql-9.6/include/server/access/genam.h:21:0,
                 from pg_btree.c:13:
/usr/pgsql-9.6/include/server/utils/relcache.h:99:13: note: declared here
 extern void RelationSetNewRelfilenode(Relation relation, char persistence,
             ^
pg_btree.c:418:23: error: dereferencing pointer to incomplete type
  if (merge || (btspool->isunique && self->max_dup_errors > 0))
                       ^
pg_btree.c:422:3: warning: implicit declaration of function '_bt_mergeload' [-Wimplicit-function-declaration]
   _bt_mergeload(self, &wstate, btspool, &reader, heapRel);
   ^
pg_btree.c:429:3: warning: implicit declaration of function '_bt_load' [-Wimplicit-function-declaration]
   _bt_load(&wstate, btspool, NULL);
   ^
pg_btree.c: At top level:
pg_btree.c:440:30: error: unknown type name 'BTWriteState'
 _bt_mergeload(Spooler *self, BTWriteState *wstate, BTSpool *btspool, BTReader *btspool2, Relation heapRel)
                              ^
pg_btree.c: In function 'BTSpoolGetNextItem':
pg_btree.c:642:38: error: dereferencing pointer to incomplete type
  return tuplesort_getindextuple(spool->sortstate, true, should_free);
                                      ^
pg_btree.c: At top level:
pg_btree.c:38:13: warning: 'unused_bt_spooldestroy' declared 'static' but never defined [-Wunused-function]
 static void unused_bt_spooldestroy(BTSpool *);
             ^
pg_btree.c:40:13: warning: 'unused_bt_spool' declared 'static' but never defined [-Wunused-function]
 static void unused_bt_spool(BTSpool *, ItemPointer self,
             ^
pg_btree.c:45:13: warning: 'unused_bt_leafbuild' declared 'static' but never defined [-Wunused-function]
 static void unused_bt_leafbuild(BTSpool *, BTSpool *);
             ^
pg_btree.c:638:1: warning: 'BTSpoolGetNextItem' defined but not used [-Wunused-function]
 BTSpoolGetNextItem(BTSpool *spool, IndexTuple itup, bool *should_free)
 ^
pg_btree.c:806:1: warning: 'BTReaderGetNextItem' defined but not used [-Wunused-function]
 BTReaderGetNextItem(BTReader *reader)
 ^
pg_btree.c:858:1: warning: 'compare_indextuple' defined but not used [-Wunused-function]
 compare_indextuple(const IndexTuple itup1, const IndexTuple itup2,
 ^
pg_btree.c:924:1: warning: 'heap_is_visible' defined but not used [-Wunused-function]
 heap_is_visible(Relation heapRel, ItemPointer htid)
 ^
pg_btree.c:938:1: warning: 'remove_duplicate' defined but not used [-Wunused-function]
 remove_duplicate(Spooler *self, Relation heap, IndexTuple itup, const char *relname)
 ^
make[2]: *** [pg_btree.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[1]: *** [all] Error 2
error: Bad exit status from /var/tmp/rpm-tmp.bRzoFr (%build)
    Bad exit status from /var/tmp/rpm-tmp.bRzoFr (%build)
make: *** [build] Error 1


```

**Comment by ooyamams** [comment]
We have already changed for 9.6. See a73dc5347b5f8b1547873fd0b8595baca5a9a013.
Can you show me pg_config?
`$ pg_config | grep VERSION`

**Comment by devrimgunduz** [comment]
I'm not building against git head -- I'm trying to build against latest stable release. Will you release a new version?

**Comment by ooyamams** [comment]
mmm... I confirmed to build for 9.6 in my environment and Travis CI is Green. So it seems to me that pg_bulkload does not have any problem.

Yes. We will release a new version and RPMs for 9.6. Please wait for a while.

**Comment by amitlan** [comment]
Build against PostgreSQL 9.6 works fine with tips of both the master and VERSION3_1_STABLE branches in the git repository.  Devrim's complaint seems to be that the build fails with the latest *released* tarball.

As @ooyamams says, a new release supporting 9.6 is on its way.

**Comment by devrimgunduz** [comment]
Thanks. Not sure about your timeline, but since 9.6 will be out tomorrow, if we miss this package, people may not upgrade. So a new release *today* would be awesome.

Regards, Devrim

**Comment by ooyamams** [comment]
We released pg_bulkload 3.1.10 now.
Thank you for the advice!
